<img src="materials/images/introduction-to-genomic-data-cover.png"/>

# Introduction to Genomics

Genomics is the branch of molecular biology concerned with the structure, function, evolution, and mapping of genomes.

In this module, you will be learning about how to process and annotate **Variant Call Format (VCF)** file using Amazon Athena. The Variant Call Format specifies the format of a text file used in bioinformatics for storing gene sequence variations.

The format has been developed with the advent of large-scale genotyping and DNA sequencing projects, such as the ***1000 Genomes Project***.

`🕒 This module should take about 30 minutes to complete.`

`✍️ This notebook is written using Python.`

<div class="alert alert-block alert-info">
<h3>⌨️ Keyboard shortcut</h3>

These common shortcut could save your time going through this notebook:
- Run the current cell: **`Enter + Shift`**.
- Add a cell above the current cell: Press **`A`**.
- Add a cell below the current cell: Press **`B`**.
- Change a code cell to markdown cell: Select the cell, and then press **`M`**.
- Delete a cell: Press **`D`** twice.

Need more help with keyboard shortcut? Press **`H`** to look it up.
</div>



---


In [1]:
# Tables you can query
# ['default.g1000vcf_csv_int', 'default.g1000vcf_csv', 'default.g1000vcf_parquet', 'default.g1000vcf_partitioned']
# COSMIC68 Annotation Dataset ['1000_genomes.hg19_cosmic68_int']
# UCSC RefGene Annotation Dataset ['1000_genomes.hg19_ucsc_refgene_int']

---

# Initial Setup

We'll use the PyAthena library to get access to a database stored in AWS S3. You can read more about PyAthena here:

• https://pypi.org/project/pyathena/

• https://aws.amazon.com/athena/?whats-new-cards.sort-by=item.additionalFields.postDateTime&whats-new-cards.sort-order=desc

In [2]:
import sys
import pyathena
import pandas as pd

from IPython import display

conn = pyathena.connect(s3_staging_dir="s3://athena-output-351869726285/", region_name='us-east-1', encryption_option='SSE_S3')

/tmp/ipykernel_101/317448135.py:3: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


---

# Query the 1000 Genomes Project Dataset

It's usually helpful to picture the data before performing any analysis, so we are going to import the database as the first step, and then view a few random rows.

<div class="alert alert-block alert-info">
<b>Tip:</b> If you are new to Jupyter Notebook, try run the code cell below using keyboard shortcut: "Shift" + "Enter". You could look up more keyboard shortcuts by pressing "H".
</div>

In [3]:
# pd.set_option('display.max_colwidth', None) # This code expands the table horizontally so that all table cells are visible.
pd.read_sql('SELECT * FROM default.g1000vcf_csv_int LIMIT 10', conn).head(10)

/tmp/ipykernel_101/653230635.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql('SELECT * FROM default.g1000vcf_csv_int LIMIT 10', conn).head(10)


chrm  start_position  end_position reference_bases alternate_bases  \
0    1        33216544      33216545               C               T   
1    1        33216657      33216658               C               T   
2    1        33233956      33233957               C               T   
3    1        33254320      33254321               C               T   
4    1        33255594      33255595               C               T   
5    1        33295904      33295905               C               T   
6    1        33335767      33335768               C               T   
7    1        33356021      33356022               C               T   
8    1        33357032      33357033               C               T   
9    1        33360652      33360653               C               T   

          rsid qual filter                                               info  \
0  rs530031665  100   PASS  AC=2;AF=0.000399361;AN=5008;NS=2504;DP=16121;E...   
1  rs565542327  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=18933;E...   
2  rs534423375  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=18999;E...   
3  rs528162115  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=16150;E...   
4  rs185431297  100   PASS  AC=6;AF=0.00119808;AN=5008;NS=2504;DP=12377;EA...   
5  rs532923873  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=18881;E...   
6  rs548587907  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=12369;E...   
7  rs144877379  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=20639;E...   
8  rs559320134  100   PASS  AC=2;AF=0.000399361;AN=5008;NS=2504;DP=18751;E...   
9  rs539478097  100   PASS  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=18273;E...   

  chromosome  
0          1  
1          1  
2          1  
3          1  
4          1  
5          1  
6          1  
7          1  
8          1  
9          1

We will go through a few important columns for you to understand in the table above. If you want to dive deeper, you may find this document helpful:https://samtools.github.io/hts-specs/VCFv4.2.pdf

1. `chrm` is the chromosome location.
2. `start_position` is the start position of the DNA variant.
3. `end_position` is the end position of the DNA variant.
4. `reference_base` is the allele at a specific location in the reference genomes, which are considered as the "approximated normal".
5. `alternate_base` is the allele that shows up at a specific location in the sample, but does not exist at the corresponding location in the reference genomes.
6. `rsid` is the identification number of the SNPs (Single Nucleotide Polymorphism), see here: https://www.snpedia.com/index.php/SNPedia
7. `qual` is the Phred-scaled quality score for the assertion made in ALT.
8. `filter` indicates PASS when the position has passed all filters.

The `info` column explains additional information about the data. In the example below, we can tell the percentage of East Asian population who has `rs559815820` SNP is only **0.0399361%**. That is 2 alleles count out of 5008 alleles from 2504 samples in the ***1000 Genomes Project***. The percentage of East Asian population who has `rs559815820` SNP is **0.1%**.

<img src="materials/images/example-1.png"/>

Now, try to reference the explanation below about the additional information section. Select a SNP from the data you queried: How does this SNP present in the ethnic group you belong to, in comparison to other ethnic groups?

1. `AC` stands for allele count in genotypes.
2. `AF` stands for allele frequency. Note: allele frequency (AF) = allele count (AC)/ allele number (AN)
3. `AN` stands for total number of alleles in genotypes.
4. `NS` stands for number of samples with data.
5. `EAS_AF` is the allele frequency of **East Asian population**.
6. `AMR_AF` is the allele frequency of **Ad Mixed American population**.
7. `EUR_AF` is the allele frequency of **European population**.
8. `AFR_AF` is the allele frequency of **African population**.
9. `SAS_AF` is the allele frequency of **South Asian population**.

---

# Search for SNPs (Single Nucleotide Polymorphism)

Next, pick a SNPs you are interested in investigating from this website: https://www.snpedia.com

The website has SNPs associated with a wide range of phenotypes. You could start exploring the popular SNPs on the homepage.

Here are a few examples:
1. `rs53576` is highly associated with the ability to empathize with others.
2. `rs72921001` is responsible for certain population's dislike towards the taste of cilantro.
3. `rs28936679` is associated with sleep disorder.
4. `rs1805009` is associated with skin cancer.

The example code below calls the data in the ***1000 Genomes Project*** that has `rs12913832`, see the code  ` WHERE rsid='rs12913832'`.

`rs12913832` is the SNP associated with blue or brown eye color.

In [4]:
pd.set_option('display.max_colwidth', None) # This code expands the table horizontally so that all table cells are visible.
pd.read_sql("SELECT * FROM \"default\".g1000vcf_csv_int WHERE rsid='rs12913832'", conn).head()

/tmp/ipykernel_101/1831946132.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql("SELECT * FROM \"default\".g1000vcf_csv_int WHERE rsid='rs12913832'", conn).head()


chrm  start_position  end_position reference_bases alternate_bases  \
0   15        28365617      28365618               A               G   

         rsid qual filter  \
0  rs12913832  100   PASS   

                                                                                                                             info  \
0  AC=888;AF=0.177316;AN=5008;NS=2504;DP=19161;EAS_AF=0.002;AMR_AF=0.2017;AFR_AF=0.028;EUR_AF=0.6362;SAS_AF=0.0706;AA=A|||;VT=SNP   

  chromosome  
0         15

Does the result above make sense to you?

We can see allele frequency of `rs12913832` among **Ad Mixed American** and **European** populations are 20.17% and 63.62% respectively, while only 0.2% among **East Asian** and 2.8% among **African** populations.

---

# Query COSMIC68 Annotation Dataset (hg19)

***COSMIC*** database is short for Catalogue of Somatic Mutations in Cancer [hg19 cosmic 68]. Learn more at
https://cancer.sanger.ac.uk/cosmic.

Now, let's take a look at few random rows of the ***COSMIC*** database:

In [5]:
pd.read_sql('SELECT * FROM "1000_genomes".hg19_cosmic68 LIMIT 10', conn).head(10)

/tmp/ipykernel_101/26311468.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql('SELECT * FROM "1000_genomes".hg19_cosmic68 LIMIT 10', conn).head(10)


chrm start_position end_position reference_bases alternate_bases  \
0   10       22209854     22209855               G               A   
1   10       22880596     22880597               G               A   
2   10       24762254     24762255               G               A   
3   10       24832945     24832946               G               A   
4   10       25877983     25877984               G               A   
5   10       26825105     26825106               G               A   
6   10       27037564     27037565               G               A   
7   10       27504517     27504518               G               A   
8   10       27702227     27702228               G               A   
9   10       27702717     27702718               G               A   

                                          cosmic_info  
0              ID=COSM917032;OCCURENCE=1(endometrium)  
1              ID=COSM1262222;OCCURENCE=1(oesophagus)  
2  ID=COSM1651216,COSM917097;OCCURENCE=1(endometrium)  
3              ID=COSM917132;OCCURENCE=1(endometrium)  
4          ID=COSM268400;OCCURENCE=1(large_intestine)  
5         ID=COSM1347396;OCCURENCE=1(large_intestine)  
6          ID=COSM295113;OCCURENCE=1(large_intestine)  
7   ID=COSM917430,COSM917429;OCCURENCE=1(endometrium)  
8                    ID=COSM119734;OCCURENCE=1(ovary)  
9         ID=COSM1347457;OCCURENCE=1(large_intestine)

Like previous datasets, we see the chromosome location, start and end positions in DNA sequence, reference allele and alternate allele.

The `cosmic_info` column tells us the types of cancer and mutation occurence.

---

# Query UCSC RefGene Annotation Dataset (hg19)

The NCBI RefSeq Genes composite track shows human protein-coding and non-protein-coding genes taken from the NCBI RNA reference sequences collection (RefSeq) [hg19 refGene]. You could learn more at the following:
- https://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/
- https://genome.ucsc.edu/cgi-bin/hgTables?db=hg19&hgta_group=genes&hgta_track=refSeqComposite&hgta_table=refGene&hgta_doSchema=describe+table+schema (Schema for NCBI RefSeq - RefSeq genes from NCBI)

<img src="materials/images/genetic-mutations.png"/>

Now, let's take a look at random rows of the dataset:

In [6]:
pd.read_sql('SELECT * FROM "1000_genomes".hg19_ucsc_refgene_int LIMIT 10', conn).head(10)

/tmp/ipykernel_101/730593987.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql('SELECT * FROM "1000_genomes".hg19_ucsc_refgene_int LIMIT 10', conn).head(10)


chrm  start_position  end_position  bin          name strand  cdsstart  \
0    X        66905782      66905968  136  NM_001348064      +  66764988   
1    X        66914514      66915917  136  NM_001348064      +  66764988   
2    X        66764464      66766604  136  NM_001348064      +  66764988   
3    X        66909397      66909600  136  NM_001348063      +  66764988   
4    X        66764464      66766604  136  NM_001348063      +  66764988   
5    X        66863097      66863249  136  NM_001348063      +  66764988   
6    X        66905851      66905968  136  NM_001348063      +  66764988   
7    X        66914514      66915917  136  NM_001348061      +  66764988   
8    X        66905851      66905968  136  NM_001348061      +  66764988   
9    X        66863097      66863249  136  NM_001348061      +  66764988   

     cdsend exoncount                            exonstarts  \
0  66905885         3           66764464,66905782,66914514,   
1  66905885         3           66764464,66905782,66914514,   
2  66905885         3           66764464,66905782,66914514,   
3  66909459         4  66764464,66863097,66905851,66909397,   
4  66909459         4  66764464,66863097,66905851,66909397,   
5  66909459         4  66764464,66863097,66905851,66909397,   
6  66909459         4  66764464,66863097,66905851,66909397,   
7  66914564         4  66764464,66863097,66905851,66914514,   
8  66914564         4  66764464,66863097,66905851,66914514,   
9  66914564         4  66764464,66863097,66905851,66914514,   

                               exonends score name2 cdsstartstat cdsendstat  \
0           66766604,66905968,66915917,     0    AR         cmpl       cmpl   
1           66766604,66905968,66915917,     0    AR         cmpl       cmpl   
2           66766604,66905968,66915917,     0    AR         cmpl       cmpl   
3  66766604,66863249,66905968,66909600,     0    AR         cmpl       cmpl   
4  66766604,66863249,66905968,66909600,     0    AR         cmpl       cmpl   
5  66766604,66863249,66905968,66909600,     0    AR         cmpl       cmpl   
6  66766604,66863249,66905968,66909600,     0    AR         cmpl       cmpl   
7  66766604,66863249,66905968,66915917,     0    AR         cmpl       cmpl   
8  66766604,66863249,66905968,66915917,     0    AR         cmpl       cmpl   
9  66766604,66863249,66905968,66915917,     0    AR         cmpl       cmpl   

  exonframes  
0    0,2,-1,  
1    0,2,-1,  
2    0,2,-1,  
3   0,2,1,1,  
4   0,2,1,1,  
5   0,2,1,1,  
6   0,2,1,1,  
7   0,2,1,1,  
8   0,2,1,1,  
9   0,2,1,1,

Here are what some of the columns mean according to Schema for NCBI RefSeq - RefSeq genes from NCBI:

1. `cdsstart`: Coding region start.
2. `cdsend`: Coding region end.
3. `exoncount`: Number of exons.
4. `strand`: + or - for strand


---

# Variant-Based Annotation

Variant-based annotation aims to look for **exact matches** between a query variant and a record in annotation datasets (i.e., two items have identical  chromosome, start position, end position, reference allele and alternative allele).

The code below uses the `JOIN` function to look for exact matches between the ***1000 Genomes Project*** dataset and the ***COSMIC*** dataset. It compares the start and end positions, reference allele, alternate allele at chromosome 2 between the two datasets.

In [7]:
pd.read_sql("SELECT A.chrm, A.start_position, A.end_position, A.reference_bases, A.alternate_bases,B.cosmic_info, A.info "
+ " FROM (SELECT * FROM \"default\".g1000vcf_csv_int WHERE chrm='2') as A "
+ " JOIN "
+ " (SELECT * FROM \"1000_genomes\".hg19_cosmic68_int WHERE chrm='2') as B "
+ " ON A.start_position=B.start_position AND A.alternate_bases=B.alternate_bases "
+ " ORDER By  A.start_position", conn).head()

/tmp/ipykernel_101/4159624003.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql("SELECT A.chrm, A.start_position, A.end_position, A.reference_bases, A.alternate_bases,B.cosmic_info, A.info "


chrm  start_position  end_position reference_bases alternate_bases  \
0    2           55184         55185               G               A   
1    2           95490         95491               C               T   
2    2          128457        128458               G               A   
3    2          133636        133637               G               A   
4    2          218857        218858               G               A   

                                                        cosmic_info  \
0                        ID=COSN206912;OCCURENCE=1(large_intestine)   
1                  ID=COSN42681;OCCURENCE=4(central_nervous_system)   
2  ID=COSN181337,COSN181637;OCCURENCE=1(stomach),2(large_intestine)   
3               ID=COSN33464;OCCURENCE=1(breast),1(large_intestine)   
4           ID=COSM1404910,COSM1404911;OCCURENCE=1(large_intestine)   

                                                                                                                     info  
0       AC=60;AF=0.0119808;AN=5008;NS=2504;DP=19415;EAS_AF=0;AMR_AF=0.0043;AFR_AF=0.0431;EUR_AF=0;SAS_AF=0;AA=G|||;VT=SNP  
1           AC=17;AF=0.00339457;AN=5008;NS=2504;DP=16854;EAS_AF=0.0169;AMR_AF=0;AFR_AF=0;EUR_AF=0;SAS_AF=0;AA=C|||;VT=SNP  
2            AC=1;AF=0.000199681;AN=5008;NS=2504;DP=17605;EAS_AF=0.001;AMR_AF=0;AFR_AF=0;EUR_AF=0;SAS_AF=0;AA=G|||;VT=SNP  
3           AC=3;AF=0.000599042;AN=5008;NS=2504;DP=19368;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0;SAS_AF=0.0031;AA=G|||;VT=SNP  
4  AC=1;AF=0.000199681;AN=5008;NS=2504;DP=17149;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0.001;SAS_AF=0;AA=G|||;VT=SNP;EX_TARGET

The returned table tells us information about genes associated with different types of cancers.

In the example below, the cancer related to central nervous system is at chromosome 2, where the reference allele is C and alternate allele is T. It appears to only happen to **East Asian** population with a frequency of 1.69% while absent among other populations.

<img src="materials/images/example-2.png"/>

---

# Interval-Based Annotation [TP53: chrm17]

The aim of interval-based annotation is to look for overlap of a query variant with a region (this region could be a single position) in annotation databases.

<img src="materials/images/overlapping-condition.png"/>

<div class="alert alert-block alert-info">
<b> Green.start_position[X]<=Blue.end_position[B] AND Blue.start_position[A]<=Green.end_position[Y]</b>

Source: https://stackoverflow.com/questions/20981783/how-to-sum-overlapping-interval-parts-in-sql.
</div>

The code below uses the `JOIN` function to compare two datasets using overlapping condition.

1. Rather than comparing whether the two datasets are exactly the same, this method focuses on overlapping regions. In the following example, we are running chromosome 17 from the ***1000 Genomes Project*** against gene TP53 from chromosome 17 in the ***UCSC RefGene Annotation*** dataset.
2. The `ON` condition is trying to see if there is any overlapping between the two datasets in the selected region. See the graph below:


In [8]:
pd.read_sql("SELECT A.chrm, A.start_position, A.end_position, A.reference_bases, A.alternate_bases,B.name, B.name2, A.info "
+ " FROM (SELECT * FROM \"default\".g1000vcf_csv_int WHERE chrm='17') as A "
+ " JOIN "
+ " (SELECT * FROM \"1000_genomes\".hg19_ucsc_refgene_int WHERE chrm='17' and name2='TP53') as B "
+ " ON A.start_position<=B.end_position AND B.start_position<=A.end_position "
+ " ORDER By  A.start_position", conn).head()

/tmp/ipykernel_101/1411853846.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql("SELECT A.chrm, A.start_position, A.end_position, A.reference_bases, A.alternate_bases,B.name, B.name2, A.info "


chrm  start_position  end_position reference_bases alternate_bases  \
0   17         7571751       7571752               T               G   
1   17         7571751       7571752               T               G   
2   17         7571751       7571752               T               G   
3   17         7571751       7571752               T               G   
4   17         7571751       7571752               T               G   

           name name2  \
0  NM_001276697  TP53   
1  NM_001126115  TP53   
2  NM_001126117  TP53   
3  NM_001276699  TP53   
4  NM_001126116  TP53   

                                                                                                            info  
0  AC=13;AF=0.00259585;AN=5008;NS=2504;DP=19133;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0.0129;SAS_AF=0;AA=T|||;VT=SNP  
1  AC=13;AF=0.00259585;AN=5008;NS=2504;DP=19133;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0.0129;SAS_AF=0;AA=T|||;VT=SNP  
2  AC=13;AF=0.00259585;AN=5008;NS=2504;DP=19133;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0.0129;SAS_AF=0;AA=T|||;VT=SNP  
3  AC=13;AF=0.00259585;AN=5008;NS=2504;DP=19133;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0.0129;SAS_AF=0;AA=T|||;VT=SNP  
4  AC=13;AF=0.00259585;AN=5008;NS=2504;DP=19133;EAS_AF=0;AMR_AF=0;AFR_AF=0;EUR_AF=0.0129;SAS_AF=0;AA=T|||;VT=SNP

<img src="materials/images/conductor.png"/>

For your information, TP53 gene provides instructions for making a protein called tumor protein p53 (or p53). This protein acts as a tumor suppressor, which means that it regulates cell division by keeping cells from growing and dividing (proliferating) too fast, or in an uncontrolled way. TP53 acts like a conductor in an orchestra.

The table gives us a view of all the mutations on gene TP53 on chromosome 2. Typically, mutations on TP53 are considered as rare mutations.

---

# Contributions & acknowledgement

Thank the following team for working on this module:

- **Module Content**: Amir Bahmani
- **Engineering**: Amit Dixit
- **UX/UI Design & Illustration**: Kexin Cha
- **Video Production**: Francesca Goncalves
- **Project Management**: Amir Bahmani, Kexin Cha

---

Copyright (c) 2022 Stanford Data Ocean (SDO)

All rights reserved.